# 문제 1.  L=[1, 2, 3, 4, 5]일 때 다음 각 문장을 수행한 후의 결과를 보고 납득할 만한 이유를 설명하시오.

In [8]:
L = [1,2,3,4,5]
L[1:3] = [100]
print(L)

[1, 100, 4, 5]


리스트는 서로다른크기에 대한 치환이 가능하다.  인덱스1과2의 원소가 100으로 치환된다.

In [10]:
L = [1,2,3,4,5]
L[1:3] = [(100)]
print(L)

[1, 100, 4, 5]


[(100)]은 [100]과 같다.  위와 같은 방식으로 치환된다.

In [36]:
L = [1,2,3,4,5]
L[1:3] = 100
print(L)

TypeError: can only assign an iterable

100은 상수이기 때문에 치환할 수 없다.

In [34]:
L = [1,2,3,4,5]
L[1:2] = (100)
print(L)

TypeError: can only assign an iterable

(100)은 100과 같으므로 상수는 리스트의 원소로 치환할 수 없다.

In [22]:
L = [1,2,3,4,5]
L[1:3] = (100, 101, 102)
print(L)

[1, 100, 101, 102, 4, 5]


튜플은 리스트의 원소로 치환할 수 있다.

In [23]:
L = [1,2,3,4,5]
L[1:3] = [100, 101, 102]
print(L)

[1, 100, 101, 102, 4, 5]


리스트는 리스트의 원소로 치환할 수 있다.

In [24]:
L = [1,2,3,4,5]
L[1:3] = [(100, 101, 102)]
print(L)

[1, (100, 101, 102), 4, 5]


리스트안에 튜플을 리스트의 원소로 치환할 때, 내부의 튜플은 그대로 유지된다.

In [25]:
L = [1,2,3,4,5]
L[1:3] = [[100, 101, 102]]
print(L)

[1, [100, 101, 102], 4, 5]


리스트안에 리스트를 리스트의 원소로 치환할 때, 내부의 리스트는 그대로 유지된다.

# 문제 2. 문자열 S = 'Hello World and Python'에 대해 다음 요구사항에 대한 Python 코드를 제시하시오.
### 1) 단어의 순서가 역순으로 되어진 문자열 ('Python and World Hello')을 만들고
### 2) 1)의 결과 문자열에 대해 메소드 split과 join을 이용하여 공백을 모두 없엔 문자열 'PythonandWorldHello'을 만드시오.

In [69]:
S = "Hello World and Python"
s = S[15:] + S[11:16] + S[6:12] + S[:5]
s = s.split()
s = ''.join(s)
print(s)

PythonandWorldHello


S를 역순으로 되어진 문자열 s로 변경하고, s.split()를 통해 공백을 기준으로 나누어 리스트로 만든다. 

만든 리스트 s를 ''.join(s)를 통해 공백없이 합친다.

# 문제 3. 다음 코드를 보고 물음에 답하시오.

##### a = [1, 2, 3]
##### b = a * 3
##### c = [a] * 3
##### print(b)
##### print(c)
##### 1) 위 코드를 수행한 이후 다음 코드를 수행한 후 b와 c의 내용을 확인하시오.
##### a[0]=0

In [73]:
a = [1, 2, 3]
b = a * 3
c = [a] * 3
print(b)
print(c)

[1, 2, 3, 1, 2, 3, 1, 2, 3]
[[1, 2, 3], [1, 2, 3], [1, 2, 3]]


In [74]:
a[0] = 0
print(b)
print(c)

[1, 2, 3, 1, 2, 3, 1, 2, 3]
[[0, 2, 3], [0, 2, 3], [0, 2, 3]]


##### 2) b와 c 객체의 차이는 무엇인가?,

b는 리스트 a를 3번 반복한 새로운 객체인 반면 c는 리스트 a를 원소로 가진다. 즉 c = [a,a,a]이므로 a의 원소가 변경되면 c의 원소도 변경된다.

# 문제 4. 다음 문자열을 ':'을 기준으로 분리하여 리스트로 만들고 각 문자열의 좌우 공백을 제거하시오 (즉, 문자열 S에서 l을 만들어라)
### s = '  first star   :   second star   :    third star  '
### l = ['first star', 'second star', 'third star']
### 1) for ~ in 문을 이용한 코드를 작성하시오.
### 2) 리스트 내포(list comprehension)를 이용한 코드를 작성하시오.

In [87]:
s = '  first star   :   second star   :    third star  '
s = s.split(':')
l = [s[i].strip() for i in range(len(s))]
print(l)

['first star', 'second star', 'third star']


문자열 s를 :기준으로 나누어 리스트로 만들고 리스트 내포와 strip()를 이용하여 좌우공백을 제거한 리스트로 만든다.

# 문제 5. 다음과 같이 0보다 큰 정수 리스트 변수 list를 인자로 받는 함수 addall(list)와 addallodd(list)를 for ~ in 리터널과 리스트 내포 방식으로 각각 제시하시오.

### 1) addall(list) 함수 (리스트 내의 모든 정수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)

##### >>> addall([1])
1
##### >>> addall([1, 2, 3, 4, 5, 6, 7, 8, 9])
45
### 2) addallodd(list) 함수 (리스트내의 모든 홀수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)
##### >>> addallodd([1])
1
##### >>> addallodd([1, 2, 3, 4, 5, 6, 7, 8, 9])
25
### [참고 1]: 리스트 내포 방식으로 addall() 및 addallodd()함수를 만들 때엔 리스트 내포에 의해 생성된 리스트 결과에 대해 내장 함수를 사용하는 방식을 고려해보시오 (주의: 리스트 내포 방식의 addall() 및 addallodd() 함수의 몸체는 단 1줄로 작성되어야 한다.)
### [참고 2]: sum() 등의 내장함수 사용 가능함 

In [109]:
def addall(a):
    return sum(i for i in a)

def addallodd(a):
    return sum(i for i in a if i%2 == 1)

a = [1,2,3,4,5,6,7,8,9,10]

print(addall(a))
print(addallodd(a))


55
25


addall함수는 내장함수 sum()을 이용하여 리스트 내포 방식으로 a의 원소를 하나씩 더해주고 
addallodd함수는 if문으로 조건을 주어 a의 원소중 홀수 인것만 더해준다. 

# 문제 6. 다음 코드를 보고 물음에 답하시오.

In [112]:
L1 = [1, 2, 3]
L2 = [4, 5, 6]
d = {'low':L1, 'high':L2}
e = d
f = d.copy()
print(d)
print(e)
print(f)
print()
d['low'] = [10, 20, 30]
d['high'][1] = 500
print(d)
print(e)
print(f)

{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}

{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [1, 2, 3], 'high': [4, 500, 6]}


### 1) 위 코드의 수행 결과 d와 e는 항상 같은 결과를 출력한다. 그 이유는 무엇인가?
e와 d는 같은 객체를 가리킨다. 따라서 d가 가리키는 객체가 변경되면 e가 가리키는 객체도 변경된다.
### 2) 위 코드의 수행 결과 마지막 f의 내용은 다음과 같다. 이 때 'high' 키의 값 내용 중 500이 출력된 이유와 'low' 키의 값 내용 중 [1, 2, 3]이 출력된 이유를 설명하시오 ([hint]: shallow copy 관점에서 설명하시오.)
f는 d의 복사본인데 d['low']는 [1,2,3]을 가리키고 있는데 d['low']가 [10,20,30]을 가리키도록 변경했기 때문에 사본인 f는 변경이 없지만
d['high'][1]은 d['high']가 가리키는 원소 [4,5,6]을 [4,100,6]으로 변경한 것이기 때문에 사본도 변경된다.

# 문제 7. 사전 d = {'one':1, 'two':2, 'three':3, 'four':4, 'five':5}가 주어졌을 때 다음 요구사항에 맞는 코드를 제시하시오
[참고]: d.keys(), d.values()를 통해 리스트를 얻어낸 후 리스트가 지니고 있는 sort(cmp=None, key=None, reverse=False)함수를 활용하시오.
##### 1) 키의 알파벳 오름차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.

In [3]:
d = {'one':1,'two':2,'three':3,'four':4,'five':5}

a = d.items()
print(sorted(a,key = None,reverse = False))


[('five', 5), ('four', 4), ('one', 1), ('three', 3), ('two', 2)]


기본값이 키값으로 되어있기 때문에 reverse = False이면 알파벳 오름차순으로 정렬한다.

##### 2) 키의 알파벳 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.

In [150]:
d = {'one':1,'two':2,'three':3,'four':4,'five':5}

a = d.items()
print(sorted(a,reverse = True))

[('two', 2), ('three', 3), ('one', 1), ('four', 4), ('five', 5)]


기본값이 키값으로 되어있기 때문에 reverse = True이면 알파벳 내림차순으로 정렬한다.

##### 3) 값의 오름차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.

In [13]:


d = {'one':1,'two':2,'three':3,'four':4,'five':5}

c = d.items()

print(sorted(d.items(), key=lambda x: x[1]))




[('one', 1), ('two', 2), ('three', 3), ('four', 4), ('five', 5)]


lambda 인자 : 리턴값을 통해 x[1]즉 값을 통해 비교한다.

##### 4) 값의 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.

In [12]:
d = {'one':1,'two':2,'three':3,'four':4,'five':5}

c = d.items()

print(sorted(d.items(), key=lambda x: x[1] ,reverse = True))

[('five', 5), ('four', 4), ('three', 3), ('two', 2), ('one', 1)]


위와 마찬가지로 값을 통해 비교를 하며 reverse = True를 통해 내림차순 정렬한다.

# 문제 8. [ACM-ICPC 문제 1673번] 치킨쿠폰

강민이는 치킨 한 마리를 주문할 수 있는 치킨 쿠폰을 n장 가지고 있다. 이 치킨집에서는 치킨을 한 마리 주문할 때마다 도장을 하나씩 찍어 주는데, 도장을 k개 모으면 치킨 쿠폰 한 장으로 교환할 수 있다.

강민이가 지금 갖고 있는 치킨 쿠폰으로 치킨을 최대 몇 마리나 먹을 수 있는지 구하여라. 단, 치킨을 주문하기 위해서는 반드시 치킨 쿠폰을 갖고 있어야 한다.

In [ ]:
try:
    while 1:
        n,k = input().split()

        n = int(n)
        k = int(k)

        c = n
        while(c>=k):
                z=int(c/k)
                c=int(c%k)
                n+=z
                c+=z

        print (n)
except:exit()

10 2
19
20 3
29
30 3
44
100 5
124


![abc](https://github.com/yugidae/2015136082/blob/master/%EC%BA%A1%EC%B2%98111.PNG?raw=true)

위의 코드를 예시를 들어 설명하면 n = 100 k = 5로 가정하자.
먼저 100개의 쿠폰으로 100마리를 시키면 100개의 도장을 얻을 수 있다.
k = 5이기 때문에 도장 5개로 치킨 1마리를 먹을 수 있다.
100개의 도장으로 20마리의 추가 치킨을 먹고 20개의 도장을 또 받는다.

20 마리의 치킨을 z값에 저장하고 기존의 n의 100값과 더하여 총 120마리가 된다.
c는 100 % 5를 의미한다 위의 예시에서는 0이된다. 즉 치킨을 시키고 남은 도장의 개수가 없다는 것이다.
c의 값에 20개의 도장의 개수를 저장한다.

while문을 통해 다시 반복하여 20마리의 치킨을 주문했을때 받은 20개의 도장을 통해 
4마리의 추가 치킨을 시킨다. 이것을 z에 저장하고 다시 n에 더하면 총 124마리가 된다.
4마리를 시키면서 4개의 도장을 받았으나 치킨 1마리시키는데 5개의 도장이 필요하기 때문에 더이상 치킨의 주문은 불가하다. 

while문을 종료시킨다.

# 문제 9. [Incremental Project] 

### 이전 Assignment 2의 마지막 문제는 웹 URL로 지정된 웹페이지를 문자열로 가져와 모든 HTML 태그를 제외한 순수 텍스트를 얻어내고 그 안에 존재하는 단어를 추출하고 단어들의 총 개수를 출력하는 프로그램을 작성하는 것이었다. 이번에는 그 마지막 숙제를 그대로 확장하여 웹 URL로 지정된 웹페이지 내 순수 텍스트 안에 존재하는 각 단어들에 대해 다음 요구사항 대로 출력하는 프로그램을 작성하시오.

##### 요구사항 1. CSS와 Javascript 내용도 모두 제거하시오 (이전 숙제에서 이미 제외하였다면 본 요구사항 무시)

##### 요구사항 2. 순수 텍스트 안에 존재하는 단어들에 대해 string 모듈을 활용하여 모든 punctuation (구두문자)를 완벽히 제거하시오.

##### 요구사항 3. 각 단어들의 출현빈도를 사전형태로 저장하여 출력하시오.

##### 요구사항 4. 다음 5개의 사이트에 대한 단어 출현빈도 결과를 출력하시오.
http://cse.koreatech.ac.kr
https://www.koreatech.ac.kr
http://www.naver.com
http://www.daum.net
http://www.nytimes.com

In [3]:
import requests

req = requests.get('http://cse.koreatech.ac.kr')

source = req.text

<!DOCTYPE html>
<html lang="ko">
<head>
<!-- META -->
<meta charset="utf-8">
<meta name="Generator" content="XpressEngine">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<!-- TITLE -->
<title>컴퓨터공학부</title>
<!-- CSS -->
<link rel="stylesheet" href="/common/css/xe.min.css?20180913173442" />
<link rel="stylesheet" href="/layouts/cse/style/prettyPhoto.css?20150805222706" />
<link rel="stylesheet" href="/layouts/cse/style/stylemobile.css?20150805222706" />
<link rel="stylesheet" href="/layouts/cse/style/mobilenavigation.css?20150805222706" />
<link rel="stylesheet" href="/modules/editor/styles/ckeditor_light/style.css?20180913173434" />
<link rel="stylesheet" href="/widgets/login_info/skins/cse_login_info_2015/default.login.css?20150420232234" />
<link rel="stylesheet" href="/widgets/content_cse/skins/cse_latest_2015/css/widget.css?20150625005943" />
<link rel="stylesheet" href="/addons/font_nanum_gothic/font-nanum_gothic.css?20150429141819" />
<link rel="stylesheet" href="/files/f

In [6]:
str = ''
count = 0

for i in range(len(source)):
    
    if (source[i] == '<' ):
        count = 1
    
        
    if ( source[i] == '>' and count == 1):
        count += 1
        continue
    if ( count == 2 ):
        str += source[i]
        
   
        

s = str.split()
print (len(s))

511


이 문제는 해결하지 못했습니다. 첫번째 요구사항 스크립트의 내용을 지울수 있는방법이 도저히 생각이 안납니다.

느낀점 : ACM-ICPC 문제 1673번은 문제 자체보다 EOF 처리를 어떻게 해야하는지 잘 몰라서 한참 걸렸던것 같다. jupyter에서는 분명히 잘 실행되는데 채첨을 하면 항상 틀리게 나와서 거의 20번가까이 틀렸던 것 같다. 다른 언어였으면 while(!=EOF)를 통해 쉽게 처리 할 수 있었는데 익숙하지 않은 python으로 하려고 하닌까 상당히 어려웠고 처음에는 무엇이 문제인지도 몰랐다. 

incremental문제는 상당히 어려워서 결국 포기했다. 첫번째 요구사항에 스크립트 내용을 제거하는것을 도저히 해결할 수가 없었다.
incremental문제의 경우 언어를 떠나서 본인의 기본적인 코딩능력이 요구되는것 같다. 